src/Orleans.Core/Messaging/ClientMessageCenter.cs

```diff
@@ -164,46 +165,63 @@ public void Stop()
 
             Utils.SafeExecute(() =>
             {
-                PendingInboundMessages.CompleteAdding();
+                PendingInboundMessages.Writer.TryComplete();
             });
 
             if (this.statisticsLevel.CollectQueueStats())
             {
                 queueTracking.OnStopExecution();
             }
             GatewayManager.Stop();
+        }
+
+        public ChannelReader<Message> GetReader(Message.Categories type)
+            => PendingInboundMessages.Reader;
 
-            foreach (var gateway in gatewayConnections.Values.ToArray())
+        public void OnReceivedMessage(Message message)
+        {
+            var handler = this.messageHandlers[(int)message.Category];
+            if (handler != null)
+            {
+                handler(message);
+            }
+            else
             {
-                gateway.Stop();
+                PendingInboundMessages.Writer.TryWrite(message);
```

Comment:
if not successfull has to be handled otherwise message is dropped

Reply:
👍 In this case I'll just log a warning. This will only happen when the client has shut down (it's an unbounded channel).